src/topn.c

```diff
@@ -0,0 +1,126 @@
+#include "data.table.h"
+
+static inline void swap(int *a, int *b) { int tmp=*a; *a=*b; *b=tmp; }
+
+static inline bool icmp(const int *x, const int a, const int b, const bool min, const bool nalast) {
+  if (x[a]==x[b]) return(a > b);
+  if (x[a]==NA_INTEGER) return(nalast);
+  if (x[b]==NA_INTEGER) return(!nalast);
+  return(min ? x[a] < x[b] : x[a] > x[b]);
+}
+static inline bool dcmp(const double *x, const int a, const int b, const bool min, const bool nalast) {
+  if (x[a]==x[b] || isnan(x[a]) && isnan(x[b])) return(a > b);
+  if (isnan(x[a])) return(nalast);
+  if (isnan(x[b])) return(!nalast);
+  return(min ? x[a] < x[b] : x[a] > x[b]);
+}
+
+static inline bool i64cmp(const int64_t *x, const int a, const int b, const bool min, const bool nalast) {
+  if (x[a]==x[b]) return(a > b);
+  if (x[a]==NA_INTEGER64) return(nalast);
+  if (x[b]==NA_INTEGER64) return(!nalast);
+  return(min ? x[a] < x[b] : x[a] > x[b]);
+}
+
+static inline bool scmp(const SEXP *restrict x, const int a, const int b, const bool min, const bool nalast) {
+  if (strcmp(CHAR(x[a]), CHAR(x[b])) == 0) return (a > b);
+  if (x[a]==NA_STRING) return(nalast);
+  if (x[b]==NA_STRING) return(!nalast);
+  return(min ? strcmp(CHAR(x[a]),CHAR(x[b]))<0 : strcmp(CHAR(x[a]),CHAR(x[b])))>0;
+}
+
+static inline bool ccmp(const Rcomplex *x, const int a, const int b, const bool min, const bool nalast) {
```

Comment:
is it possible to do `dcmp()` on the `r`/`i` components here? I suspect no because of recent changes to how missingness/`NaN` works for complex values...

Reply:
We could make it work but it wouldn't be simpler than this approach.